### PR TITLE
DCM 59: Implement export DHIS mappings with underlying metadata feature

### PR DESCRIPTION
**Pull request by Gravewalker666**
### Purpose

fix the ticket [[DCM-59] Implement export DHIS mapping files with the underlying metadata feature](https://issues.openmrs.org/browse/DCM-59)

### Approach

Include related metadata as XML files into the export bundle
Write unit tests to test mapping sharing
Create a new file to keep utility functions of the DHISService file

### Environment

- Java 8 Oracle
- Maven 3.6.3
- OpenMRS Reference Application 2.11.0
- MySQL server
- macOS bigSur

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java [hunk 1]
```diff
@@ -917,102 +935,135 @@ public String uploadMappings(MultipartFile mapping) {
 		
 		return msg;
 	}
-	
+
 	@Override
-	public String[] exportSelectedMappings(String[] selectedMappings) {
-		String[] cleanedSelectedMappings = cleanSelectedMappings(selectedMappings);
+	public String[] exportMappings(String[] mappings) throws IOException {
```
Comment (jayasanka-sack) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:940]: This method is a bit lengthy. Shall we break this into sub methods to make it more readable?

File: api/src/test/java/org/openmrs/module/dhisconnector/api/DHISConnectorServiceTest.java [hunk 2]
```diff
@@ -154,4 +164,26 @@ public void testDeleteLocationToOrgUnitMappingsByLocation() {
...
+
+	@Test
+	public void testDHISMappingExport() throws IOException {
+		DHISConnectorService dhisConnectorService = Context.getService(DHISConnectorService.class);
+		ReportDefinitionService reportDefinitionService = Context.getService(ReportDefinitionService.class);
+
+		DHISMapping mapping = new DHISMapping();
+		mapping.setName("mapping-test-unit");
+		mapping.setCreated(new Date().getTime());
+		PeriodIndicatorReportDefinition pir = new PeriodIndicatorReportDefinition();
```
Comment (jayasanka-sack) [inline api/src/test/java/org/openmrs/module/dhisconnector/api/DHISConnectorServiceTest.java:176]: How about replacing `pir` with `reportDefinition` or something?

**Review by jayasanka-sack — COMMENTED** [review]
Thanks for getting this done, @Gravewalker666 ! I added some comments. Could you please check them?
Just some readability issues.

**Comment by dkayiwa** [comment]
With only the changes in this pull request, did you compile the module, install it, and confirm that the export still works?

File: api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            zOut.putNextEntry(zipEntry);
+
+            byte[] bytes = new byte[1024];
+            int length;
+            while ((length = fis.read(bytes)) >= 0) {
+                zOut.write(bytes, 0 , length);
+            }
+            fis.close();
+        }
+        zOut.close();
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java:35]: Are these streams closed when an exception is thrown?

File: api/src/test/java/org/openmrs/module/dhisconnector/api/DHISConnectorServiceTest.java (hunk 2 above)
Comment (Gravewalker666) [reply api/src/test/java/org/openmrs/module/dhisconnector/api/DHISConnectorServiceTest.java:176]: Sure @jayasanka-sack On it

File: api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java (hunk 3 above)
Comment (Gravewalker666) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java:35]: @dkayiwa 
Oh yeah it slipped my mind, I'll use a try, catch, finally block to close the stream, thank you

File: api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java (hunk 1 above)
Comment (Gravewalker666) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:940]: I'll see what I can do but it's a bit hard to break this down

**Comment by Gravewalker666** [comment]
> With only the changes in this pull request, did you compile the module, install it, and confirm that the export still works?

@dkayiwa it does work fine, but when you change the scope of XML mapper to `provided` it throws a `classNotFound` exception. I even tried adding the jar into my local maven repository to make it available at run time but no luck.

I tested the implementation with the default scope and it works like a charm.

**Comment by dkayiwa** [comment]
> > With only the changes in this pull request, did you compile the module, install it, and confirm that the export still works?
> 
> @dkayiwa it does work fine, but when you change the scope of XML mapper to `provided` it throws a `classNotFound` exception. I even tried adding the jar into my local maven repository to make it available at run time but no luck.
> 
> I tested the implementation with the default scope and it works like a charm.

Oh i see! You can restore it to what works.

**Comment by Gravewalker666** [comment]
> > > With only the changes in this pull request, did you compile the module, install it, and confirm that the export still works?
> > 
> > 
> > @dkayiwa it does work fine, but when you change the scope of XML mapper to `provided` it throws a `classNotFound` exception. I even tried adding the jar into my local maven repository to make it available at run time but no luck.
> > I tested the implementation with the default scope and it works like a charm.
> 
> Oh I see! You can restore it to what works.

Sure, I updated the PR

File: api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java (hunk 1 above)
Comment (Gravewalker666) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:940]: @jayasanka-sack i broke it down, please review

Comment (jayasanka-sack) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:940]: Looks good to me!

**Review by jayasanka-sack — APPROVED** [review]
Thanks for getting this done @Gravewalker666 !

**Pull request merged**